### PR TITLE
Alias BaseOperator.__matmul__ to BaseOperator.dot

### LINKED ISSUE #7502: Overloading `@` for `Operator` class
## Proposition

Overloading the `@` operator for class `qiskit.quantum_info.operators.Operator` ([documentation](https://qiskit.org/documentation/stubs/qiskit.quantum_info.Operator.html?highlight=operator#qiskit.quantum_info.Operator)) so it works the same as `Operator.compose()` method. The same behavior is done with class `qiskit.opflow.operator_base.OperatorBase` ([documentation](https://qiskit.org/documentation/stubs/qiskit.opflow.OperatorBase.html#qiskit.opflow.OperatorBase)).

## Motivation

When going through the Open Science Prize 2021 I encountered [problem](https://github.com/qiskit-community/open-science-prize-2021/issues/5). After investigation it turned out that `Operator` class doesn't have `@` overloading, so in order to fix the issue `compose()` or `np.matmul()` methods need to be used. As this is good solution it would be easier to use `@`, especially for reading the code.

### PR DESCRIPTION
**Pull request by daiki0623**
### Summary

Adds a `@` (`__matmul__`) binary operator. (fixes #7502) 

### Details and comments

Adds a `@` (`__matmul__`) binary operator to `BaseOperator` subclasses
in the qiskit.quantum_info module. This is shorthand to call the
classes `dot` method ( `A@ B == A.dot(B)`).

### CODE REVIEW

**Review by ikkoham — CHANGES_REQUESTED** [review]
For the test cases, it would be better to add a new test instead of replacing `&` with `@`.

**Comment by daiki0623** [comment]
@ikkoham Thank you, fixed test (add assertions to "test_dot")

File: qiskit/quantum_info/operators/measures.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def process_fidelity(channel, target=None, require_cp=True, require_tp=True):
     if isinstance(target, Operator):
         # Compute fidelity with unitary target by applying the inverse
         # to channel and computing fidelity with the identity
-        channel = channel.compose(target.adjoint())
+        channel = channel @ target.adjoint()
```
Comment (jakelishman) [inline qiskit/quantum_info/operators/measures.py:118]: This looks like a behavioural change - `compose` and `dot` aren't the same.  I'm not sure which is the correct way round here (and I'm not keen that we don't have a test failure if the behaviour changes), but the `compose` form feels more correct to me.

Comment (daiki0623) [reply qiskit/quantum_info/operators/measures.py:118]: Thnak you, you're right. I changed it by mistake. reset my change.

**Review by jakelishman — APPROVED** [review]
Thanks for the change, looks good to me!

**Comment by ikkoham** [comment]
@daiki0623 Thanks. Could you add more tests about test_sparse_pauli_op and test_scalar_op?

**Review by ikkoham — APPROVED** [review]
Thank you. LGTM:rocket:

**Review by jakelishman — APPROVED** [review]

**Pull request merged**